doctr/models/classification/vip/layers/pytorch.py

```diff
@@ -0,0 +1,615 @@
+
+        self.mixer = Attention(
+            dim,
+            num_heads=num_heads,
+            qkv_bias=qkv_bias,
+        )
+
+        self.drop_path = DropPath(drop_path_rate) if drop_path_rate > 0.0 else nn.Identity()
+        self.norm2 = nn.LayerNorm(dim)
+        mlp_hidden_dim = int(dim * mlp_ratio)
+        self.mlp = PositionwiseFeedForward(d_model=dim, ffd=mlp_hidden_dim, dropout=0.0, activation_fct=nn.GELU())
+
+    def forward(self, x: torch.Tensor, size: tuple[int, int] | None = None) -> torch.Tensor:
+        """
+        Forward pass for MultiHeadSelfAttention.
+
+        Args:
+            x: A float tensor of shape (b, n, c).
+            size: An optional (h, w) if needed by some modules (unused here).
+
+        Returns:
+            A float tensor of shape (b, n, c) after self-attention and MLP.
+        """
+        x = x + self.drop_path(self.mixer(self.norm1(x)))
+        x = x + self.drop_path(self.mlp(self.norm2(x)))
+        return x
+
+
+class OverlappedSpatialReductionAttention(nn.Module):
+    """
+    Overlapped Spatial Reduction Attention (OSRA).
+
+    This attention mechanism downsamples the input according to 'sr_ratio' (spatial reduction ratio),
+    applies a local convolution for feature enhancement. It captures dependencies in an overlapping manner.
+
+    Args:
+        dim: The embedding dimension of the tokens.
+        num_heads: Number of attention heads.
+        qk_scale: Optionally override q-k scaling. Defaults to head_dim^-0.5 if None.
+        attn_drop: Dropout rate for attention weights.
+        sr_ratio: Spatial reduction ratio. If > 1, a depthwise conv-based downsampling is applied.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        num_heads: int = 1,
+        qk_scale: float | None = None,
+        attn_drop: float = 0.0,
+        sr_ratio: int = 1,
+    ) -> None:
+        super().__init__()
+        assert dim % num_heads == 0, f"dim {dim} should be divisible by num_heads {num_heads}."
+        self.dim = dim
+        self.num_heads = num_heads
+        head_dim = dim // num_heads
+        self.scale = qk_scale or head_dim**-0.5
+        self.sr_ratio = sr_ratio
+        self.q = nn.Conv2d(dim, dim, kernel_size=1)
+        self.kv = nn.Conv2d(dim, dim * 2, kernel_size=1)
+        self.attn_drop = nn.Dropout(attn_drop)
+
+        if sr_ratio > 1:
+            self.sr = nn.Sequential(
+                *conv_sequence_pt(
+                    dim,
+                    dim,
+                    kernel_size=sr_ratio + 3,
+                    stride=sr_ratio,
+                    padding=(sr_ratio + 3) // 2,
+                    groups=dim,
+                    bias=False,
+                    bn=True,
+                    relu=False,
+                ),
+                nn.GELU(),
+                *conv_sequence_pt(dim, dim, kernel_size=1, groups=dim, bias=False, bn=True, relu=False),
+            )
+        else:
+            self.sr = nn.Identity()
```

Comment:
mypy ignore: `self.sr = nn.Identity()  # type: ignore[assignment]`

or in line 233 `self.sr: nn.Sequential | nn.Identity`